wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/environment/server/CheEnvironmentEngine.java

```diff
@@ -0,0 +1,954 @@
+        EnvironmentHolder environmentHolderCopy;
+        try (StripedLocks.ReadLock lock = stripedLocks.acquireReadLock(workspaceId)) {
+            EnvironmentHolder environmentHolder = environments.get(workspaceId);
+            if (environmentHolder == null || environmentHolder.status != EnvStatus.RUNNING) {
+                throw new ConflictException(format("Environment '%s' is not running", workspaceId));
+            }
+            environmentHolderCopy = new EnvironmentHolder(environmentHolder);
+        }
+        for (Instance machine : environmentHolderCopy.machines) {
+            if (machine.getConfig().getName().equals(machineConfig.getName())) {
+                throw new ConflictException(
+                        format("Machine with name '%s' already exists in environment of workspace '%s'",
+                               machineConfig.getName(), workspaceId));
+            }
+        }
+        String machineId = generateMachineId();
+        final String creator = EnvironmentContext.getCurrent().getSubject().getUserId();
+
+        Instance instance = null;
+        try {
+            addMachine(workspaceId,
+                       machineId,
+                       environmentHolderCopy.name,
+                       creator,
+                       machineConfig);
+
+            eventService.publish(newDto(MachineStatusEvent.class)
+                                         .withEventType(MachineStatusEvent.EventType.CREATING)
+                                         .withDev(machineConfig.isDev())
+                                         .withMachineName(machineConfig.getName())
+                                         .withMachineId(machineId)
+                                         .withWorkspaceId(workspaceId));
+
+            instance = startMachineInstance(machineConfig,
+                                            workspaceId,
+                                            machineId,
+                                            environmentHolderCopy.name,
+                                            creator,
+                                            false,
+                                            environmentHolderCopy.logger);
+
+            replaceMachine(instance);
+
+            eventService.publish(newDto(MachineStatusEvent.class)
+                                         .withEventType(MachineStatusEvent.EventType.RUNNING)
+                                         .withDev(machineConfig.isDev())
+                                         .withMachineName(machineConfig.getName())
+                                         .withMachineId(instance.getId())
+                                         .withWorkspaceId(workspaceId));
+
+            return instance;
+        } catch (Exception e) {
+            removeMachine(workspaceId, machineId);
+
+            if (instance != null) {
+                try {
+                    instance.destroy();
+                } catch (Exception ignore) {
```

Comment:
Why is it ignored ?
It looks like it shouldn't normally happen, so it should be  at least logged.